frontend/lib/uast/post-parse-checks.cpp

```diff
@@ -223,6 +224,15 @@ static ControlFlowModifier nodeAllowsReturn(const AstNode* node,
       // can't return a value from an iterator.
       return ControlFlowModifier::BLOCKS;
     }
+
+    if(fn->name() == USTR("deinit") ||
```

Comment:
Can we return values from `init` functions?

Reply:
Typically no as demonstrated by this this existing test that tries to return from an `init` function:

https://github.com/stonea/chapel/blob/special_method_kwords_exp/test/classes/initializers/noReturnInInit.chpl#L88

And it's `.good` locking down an error:

https://github.com/stonea/chapel/blob/special_method_kwords_exp/test/classes/initializers/noReturnInInit.good

Because of existing error handling code in [initializerRules.cpp](https://github.com/chapel-lang/chapel/blob/60c2e02d2667c584f84356d748c66b3ae0daf000/compiler/passes/initializerRules.cpp#L234)

If I go ahead and add `init` to my new checks in `post-parse-checks.cpp` (in addition to getting failures due to rewording the error message) I get some (somewhat unexpected) failures that deal with `lifetime return`.

This [test demonstrates a case of that](https://github.com/stonea/chapel/blob/special_method_kwords_exp/test/classes/delete-free/lifetimes/lifetime-clause-init.chpl) gets captured as well.  as well as the use of `lifetime return` on the [init function for ArgumentParser as well](https://github.com/chapel-lang/chapel/blob/main/modules/packages/ArgumentParser.chpl#L969C3-L969C3).

Reply:
I see, so it's disallowed elsewhere already, but if you try to catch it earlier here we get some strange issues. (I guess we use `lifetime return` on an `init` to declare the lifetime of the thing getting `init`ed, and there's some point between here and `initializerRules` that treats it as a return?) I think it's ok to not try and debug this now, but I suggest a comment here that says it's caught later.

Reply:
> I see, so it's disallowed elsewhere already, but if you try to catch it earlier here we get some strange issues.

That's right.  Arguably, maybe I should do the checks for `postinit` and `deinit` in `initializerRules.cpp` as well just so they're closer to each other.  Though again, maybe it would be better to move some of those checks from `initializerRules` into `post-parse-checks` (getting errors earlier is better, right?)?  I'm really not sure, so I'm a bit hesitant to start refactoring it especially without having a better understanding of why `initializerRules` is its own separate pass.

> and there's some point between here and initializerRules that treats it as a return?

I'd say it's treated as a return coming out of Dyno. In `Visitor::visit(const Return* node)` in `post-parse-checks` it gets picked the return up and when we call `node->toFunction()` on it that it returns that it's in an `init` method.  I suppose at this point there's probably some more advanced pattern matching I could do to tell whether the `return` is a "real" return or a return for the `lifetime` annotation.  But again, this is already being handled later down in the compilation so why repeat it here?

I can definitely add a comment explaining that we'll check for the `init` case later on, cause I agree it does seem out-of-place to not see it alongside `deinit` and `postinit`.

Reply:
That course of action sounds good to me

Reply:
Ok, I added this comment:

```
      // The 'init' method is handled separately by the initializerRules pass.
      // If we handle it here it also erroneously picks up the use of 'return'
      // in 'lifetime return' statements (which isn't a concern for deinit
      // and postinit).
```